src/Disciplina/Educator/Validate.hs

```diff
@@ -0,0 +1,41 @@
+
+-- | Validation of private block
+
+module Disciplina.Educator.Validate
+       ( validate
+       ) where
+
+import Universum
+
+import Control.Monad.Except (MonadError)
+import Serokell.Util.Verify (verifyGeneric, verResToMonadError)
+
+import Disciplina.Educator.Block (PrivateBlock (..), PrivateBlockHeader (..), pbBody, pbbTxs)
+import Disciplina.Educator.Txs (PrivateTx (..), PrivateTxAux (..), PrivateTxWitness (..))
+import Disciplina.Educator.Serialise ()
+import Disciplina.Core (Submission (..), SignedSubmission (..), Address (..))
+import Disciplina.Crypto (fromFoldable, getMerkleRoot, hash, verify, )
+
+-- | Validate private block.
+-- Block is valid if
+-- 1. Header matches body
+-- 2. All transactions have valid signatures
+-- 3. Transaction signature public key correspond to student public key
+validate :: (MonadError [Text] m) => PrivateBlock -> m ()
+validate pb =
+    let txs = pb ^. pbBody . pbbTxs
+        merkleRoot = getMerkleRoot (fromFoldable txs)
+        headerVer = validateTxHeader (_pbHeader pb) merkleRoot
+        blockValid = verifyGeneric (headerVer <> concatMap validateTx txs)
+    in verResToMonadError toList blockValid
+  where validateTxHeader (PrivateBlockHeader {..}) merkleRoot =
+          [(_pbhBodyProof == merkleRoot, "Header signature missmatch")]
+
+        validateTx (PrivateTxAux {..}) =
+          let txProofSig = _ptwSig _ptaWitness
+              txProofKey = _ptwKey _ptaWitness
+              tx = hash _ptaTx
+              txAddrHash = addrHash (sStudentId (ssSubmission (_ptxSignedSubmission _ptaTx)))
+          in [(verify txProofKey tx txProofSig, "Tx signature not valid")
+             ,(hash txProofKey == txAddrHash, "Tx public key missmatch")
```

Comment:
Typo: should be "mismatch".

Reply:
Actually, it's way better to define a custom type for tx validation and block validation errors and define `Show`/`Buildable` instances for it to get proper error messages. This way you also wouldn't need to compare string literals in tests (which is ugly).

Reply:
Sounds good. Could you elaborate on how to do this?